aten/src/ATen/native/native_functions.yaml

```diff
@@ -1477,11 +1477,7 @@
 - func: smm(Tensor self, Tensor mat2) -> Tensor
   variants: function, method
 
-# FIXME: These could be combined as optional<ScalarType> but for https://github.com/pytorch/pytorch/issues/6593.
-- func: softmax(Tensor self, int64_t dim, ScalarType dtype) -> Tensor
-  variants: function, method
-
-- func: softmax(Tensor self, int64_t dim) -> Tensor
+- func: softmax(Tensor self, int64_t dim, *, ScalarType? dtype=None) -> Tensor
```

Comment:
For `softmax` and `log_softmax`, why do we want to make it a kw only argument instead of simply merging them together? It is indeed a BC change and I don't know if we want to do that or not.

Reply:
It's mainly for the sake of consistency with other ops.

Some pro arguments are:
- These two are undocumented functions. The public interface is through `torch.nn.functional.[log_]softmax` which already have `dtype` as a kwarg.
- kwarg and non-kwarg forms are the same cpp signature so it's tricky to make them live side-by-side for a deprecation period.
- If this change is ever to be made, it's better to do it sooner than later.
- It's a better baseline ever if more kwargs are added across multiple ops (`casting`, `where` etc.).
- The fix is trivial.

Reply:
sounds good, as long as it only expose from nn.functional I think would be fine :)